packages/backend/src/controllers/authentication.ts

```diff
@@ -340,24 +340,26 @@ export const storeOIDCRedirect: RequestHandler = (req, res, next) => {
 export const getSuccessURLWithReturnTo = (req: Request): string => {
     const url = new URL('/api/v1/oauth/success', lightdashConfig.siteUrl);
     if (req.session.oauth?.returnTo) {
-        url.searchParams.set('returnTo', req.session.oauth.returnTo);
+        url.searchParams.set(
+            'returnTo',
+            decodeURIComponent(req.session.oauth.returnTo),
+        );
     }
     return url.href;
 };
 
 export const redirectOIDC: RequestHandler = (req, res) => {
     // Workaround for https://github.com/jaredhanson/passport/pull/941
-    const queryReturn =
-        typeof req.query.returnTo === 'string'
-            ? decodeURIComponent(req.query.returnTo.toString())
-            : undefined;
-
+    const queryReturn = req.query.returnTo;
     if (
         queryReturn &&
+        typeof queryReturn === 'string' &&
```

Comment:
let's parse this just once, so that we operate on the validated value instead of reconstructing the url:
```typescript
...
let url
try {
  url = new URL(queryReturn, ...)
  // do stuff with url
}
catch (e) {
  if (e.name == 'ERR_INVALID_URL') {
    res.redirect('/')
  }
}